### ISSUE TITLE
Q: Extra mode for the `end`

### ISSUE THREAD

**Issue by tenbits** [isagalaev/highlight.js#955]
Hi Team,
how it is possible to define the ending mode, similar to the `starts`:
```
begin: /~\[/, 
className: "subst smartquote",
starts: {
  end: /\]/,
  subLanguage: 'javascript',              
  excludeEnd: true,
  className: 'subst',
}
```
This should be the interpolation quotes within string mode and with javascript mode inside. Like this:
```mask
'Hello ~[foo + 2] bar'
```
Now the problem I have, the end quote `]` does not belong to `"subst smartquote"` mode, but to the parent _(`"string"`)_. And when I set `excludeEnd: true,` to `false`, the `]` belongs to javascript mode. How I can highlight the end matching?

Thanks in advance!

**Comment by isagalaev** [comment]
From my understanding of the problem something as simple as this should work:

```js
{
  className: 'subst',
  begin: /~\[/, end: /\]/,
  subLanguage: 'javascript'
}
```

**Comment by tenbits** [comment]
Thanks Ivan, but how I put the interpolation quotes _(the beginning match and the ending)_ to the seperate mode, for instance the `"smartquote"`, but not to the javascript content itself.

And how I make the content to be ungreedy, I thought actually the `starts` feature is for this, sothat the sublanguage mode is ended when we are in the root mode of that language **and** the ending is matched, like in TextMate.

So that such things would work:
```mask
"Be ~[ users[0].name ] good"
```

When we use `begin: /~\[/, end: /\]/,`, then we will stop the interpolation after `users[0]`.

**Comment by isagalaev** [comment]
Oh, that's a whole different can of worms then :-(

Unfortunately `starts` won't help here, it starts a new mode unconditionally but that mode is just a regular one, it still needs `end` or `endsWithParent` to be closed. A sub-language doesn't affect parsing of the outer language at all, it's actually applied during a second pass when the whole outer mode is parsed up to the next terminator. Which means that dealing with nested brackets and a sub language is impossible as every part between `[ ... [ ... ] ... ]` will be treated separately.

To make it work we'll need to rework this logic so a sub-language syntax tree would actually be a part of the outer language, and it's going to be quite an undertaking!

**Comment by tenbits** [comment]
Thank you, Ivan, I appreciate your explanation. And such sub-language tree nesting would be much desired. As for now, I'm trying to create the syntax highlighting for a template markup, which accepts not only the javascript interpolations, but also javascript and css style blocks. For instance:
```mask
header {
  style scoped {
       div.panel { background-image: url ('foo.png') }
       /* css comments */
       a { text-decoration: underline }
  }
  // ....
}
```
So as I understood, it is not possible as for now, to select the whole content inside `style scoped {}` block as `css` sub-language, while we have to select the block with regexp first, and it is not possible to select recursively the content in curly braces.

**Comment by isagalaev** [comment]
Hello again! It's been a while, but after our recent big release I've got a chance to look over this issue again and I think I can offer a solution here:

```js
/*
Language: my-template
Requires: css.js
*/

function (hljs) {
  return {
    contains: [
      hljs.C_LINE_COMMENT_MODE,
      {
        begin: /style\s+scoped\s*\{/, end: /\}/,
        keywords: 'style scoped',
        contains: hljs.getLanguage('css').contains
      }
    ]
  }
}
```

By using `Requires:` in the language header you can ensure that the language being embedded is compiled into the `hljs` object by this moment and then you can simply reuse its whole structure in your `contains`.

There might be some corner case bugs that I'm not aware of, but the solution itself seems sound to me, and it works perfectly with the snippet you've provided earlier.

**Comment by isagalaev** [comment]
Or even this way:

```js
contains: [
  hljs.C_LINE_COMMENT_MODE,
  hljs.inherit(hljs.getLanguage('css'), {
    begin: /style\s+scoped\s*\{/, end: /\}/,
    keywords: 'style scoped'
  })
]
```

`hljs.inherit` copies over the entire language mode, so you get not only `contains` but all the other top-level stuff like keywords, illegals, etc.

**Comment by tenbits** [comment]
Woow, Ivan, that's awesome!! Thanks, I have tested a little bit, and it works) Later I will make the deeper check.

**Comment by tenbits** [comment]
Hi Ivan, sorry for bothering you again, but it seems something wrong with the recursion and the scope in javascript syntax. 
```javascript
{
  className: 'string',
  begin: '"', 
  end: '"',        
  contains: [ 
    hljs.inherit(hljs.getLanguage('javascript'), {
      begin: /~\[/, end: /\]/,
      className: 'subst'
    })
  ]
}
```
The string interpolation works here just fine, until we have a bracket notation inside:
```mask
" Foo ~[ baz['qux'] + 2]  "
```
Then only `~[ baz['qux']` is highlighted as `subst`, and ` + 2]` belongs to the `string` scope. 
But the `css` include from the example above works as expected - `}` closes the `css` block,  when being only in the most top level scope.

Have you any ideas what can be here with `javascript` syntax wrong?
Thank you in advance!

**Comment by tenbits** [comment]
It seems `javascript` doesn't handle the brackets `[]`, the above `javascript` interpolation works, when I add to the `javascript`s syntax simple snippet inside
```javascript
{
    begin: /\[/, end: /\]/,
}
```
Do you know any solutions, how to solve this without changing the `javascript` syntax, or would it be OK to add those changes to the `javascript` syntax?

**Comment by isagalaev** [comment]
Something like this then:

```js
var js = hljs.getLanguage('javascript');

{
  className: 'string',
  begin: '"', 
  end: '"',        
  contains: [ 
    {
      className: 'subst',
      begin: /~\[/, end: /\]/,
      illegal: js.illegal,
      keywords: js.keywords,
      contains: [
        {begin: /\[/, end: /\]/, relevance: 0, contains: ['self']}
      ].concat(js.contains)
    }
  ]
}
```

I had to give up `hljs.inherit` and assign individual properties because otherwise `contains` would've been a reference to the javascript's own contains, not a copy, and modifying it would've affected pure JavaScript as well.

**Comment by tenbits** [comment]
Hi Ivan, thanks a lot for your support with this. Returning to the initial subject - how it would be possible to define the extra mod for the ending match only. Something like this would be perfect
```javascript
    {
      className: 'subst',
      begin: /~\[/, end: /\]/,
      beginClassName: 'meta', endClassName: 'meta',
      contains: [ /*.. */]
    }
```
I was trying to achieve this with such a workaround:
```javascript
begin: /~\[/, 
className: 'meta',
starts: {
  contains: [
    /* js Modes */
    {
      begin: /\]/,
      endsParent: true,
      className: 'meta'
    }
  ]
}
```
But having an infinite loop by matching.

Cheers, Alex

**Comment by tenbits** [comment]
OK, it looks like I have managed to find out a woraround to achieve the issues subject.

```javascript
function createBlock(def) {
  return {
    begin: def.begin, end: def.end,
    className: def.className,
    endsParent: def.endsParent,
    returnBegin: true,
    contains: [
      {
        begin: def.begin,
        className: def.beginClassName,
      },
      {
        begin: def.end,
        className: def.endClassName,
        endsParent: true,
      },
      {
        end: def.end,
        returnEnd: true,
        contains: def.contains
      }
    ]
  };
}
```

Usage sample:
```javascript
/*
Language: my-template
Requires: javascript.js
*/

function (hljs) {
  return {
    contains: [
      createBlock({
        begin: /~\[/, 
        end: /\]/, 
        beginClassName: 'meta',        
        endClassName: 'meta',
        contains: hljs.getLanguage('javascript').contains, 
      })
    ]
  }
}
```

_@isagalaev If you find this solution appropriate for usage, please close the issue as it is resolved then, and if I find later some issues with the solution I will reopen the issue. 
And thank you for your support_  :smiley:

**Comment by isagalaev** [comment]
Closing this in favor of the pull request using all these ideas — #1037. It's better to work on that one.
